src/main/kotlin/org/rust/ide/refactoring/move/common/RsMoveConflictsDetector.kt

```diff
@@ -0,0 +1,159 @@
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.refactoring.move.common
+
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.parentsWithSelf
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import com.intellij.refactoring.util.RefactoringUIUtil
+import com.intellij.util.containers.MultiMap
+import org.rust.ide.refactoring.move.common.RsMoveUtil.addInner
+import org.rust.ide.refactoring.move.common.RsMoveUtil.isInsideMovedElements
+import org.rust.ide.refactoring.move.common.RsMoveUtil.isSimplePath
+import org.rust.ide.refactoring.move.common.RsMoveUtil.startsWithSelf
+import org.rust.ide.utils.getTopmostParentInside
+import org.rust.lang.core.macros.setContext
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+class RsMoveConflictsDetector(
+    private val conflicts: MultiMap<PsiElement, String>,
+    private val elementsToMove: List<ElementToMove>,
+    private val sourceMod: RsMod,
+    private val targetMod: RsMod
+) {
+
+    fun detectInsideReferencesVisibilityProblems(insideReferences: List<RsMoveReferenceInfo>) {
+        for (reference in insideReferences) {
+            val pathOld = reference.pathOld
+            val target = reference.target
+            if (reference.pathNewAccessible == null) {
+                addVisibilityConflict(conflicts, pathOld, target)
+            }
+        }
+
+        detectPrivateFieldOrMethodInsideReferences()
+    }
+
+    fun detectOutsideReferencesVisibilityProblems(outsideReferences: List<RsMoveReferenceInfo>) {
+        for (reference in outsideReferences) {
+            if (reference.pathNewAccessible == null) {
+                addVisibilityConflict(conflicts, reference.pathOld, reference.target)
+            }
+        }
+
+        detectPrivateFieldOrMethodOutsideReferences()
+    }
+
+    /**
+     * For now we check only references from [sourceMod],
+     * because to check all references we should find them,
+     * and it would very slow when moving e.g. many files.
+     */
+    private fun detectPrivateFieldOrMethodInsideReferences() {
+        val movedElementsShallowDescendants = movedElementsShallowDescendantsOfType<RsElement>(elementsToMove)
+        val sourceFile = sourceMod.containingFile
+        val elementsToCheck = sourceFile.descendantsOfType<RsElement>() - movedElementsShallowDescendants
+        detectPrivateFieldOrMethodReferences(elementsToCheck.asSequence(), this::checkVisibilityForInsideReference)
+    }
+
+    /** We create temp child mod of [targetMod] and copy [target] to this temp mod */
+    private fun checkVisibilityForInsideReference(referenceElement: RsElement, target: RsVisible) {
+        if (target.visibility == RsVisibility.Public) return
+        // can't use `containingMod` directly,
+        // because if `target` is expanded by macros, then it will be inside special file
+        val targetContainingMod = target.ancestorStrict<RsMod>() ?: return
```

Comment:
@vlad20012 is this a proper way to get potentially expanded containing mod? I supposed `containingMod` is quite enough

Reply:
We are not interested in expanded elements here, and `getTopmostParentInside` will fail if we used `containingMod` when `target` is expanded by macros. Also please see updated comment

Reply:
Oh, I misunderstood the old version of this comment, thank you for the explanation :)